arangod/Cluster/SynchronizeShard.cpp

```diff
@@ -874,59 +749,45 @@ bool SynchronizeShard::first() {
-    // protocol. this shortcut relies on the collection counts being correct, and
-    // as these can be at least temporarily off, we need to disable it.
-#if 0
-    if (docCount == 0) {
-      // We have a short cut:
-      LOG_TOPIC("0932a", DEBUG, Logger::MAINTENANCE)
-          << "synchronizeOneShard: trying short cut to synchronize local shard "
-             "'"
-          << database << "/" << shard << "' for central '" << database << "/"
-          << planId << "'";
-
-      // now do a final sync-to-disk call. note that this can fail
-      Result res = EngineSelectorFeature::ENGINE->flushWal(/*waitForSync*/ true, /*waitForCollector*/ false);
-      if (res.fail()) {
-        LOG_TOPIC("a49d1", INFO, Logger::MAINTENANCE) << res.errorMessage();
-        result(res);
-        return false;
-      }
+    LOG_TOPIC("53337", DEBUG, Logger::MAINTENANCE)
+        << "synchronizeOneShard: trying to synchronize local shard '" << database
+        << "/" << shard << "' for central '" << database << "/" << planId << "'";
+    
+    std::shared_ptr<DatabaseTailingSyncer> tailingSyncer;
 
-      try {
-        NetworkFeature& nf = _feature.server().getFeature<NetworkFeature>();
-        network::ConnectionPool* pool = nf.pool();
-        auto asResult = addShardFollower(pool, ep, database, shard, 0, clientId,
-                                         SyncerId{}, _clientInfoString, 60.0);
-
-        if (asResult.ok()) {
-          if (Logger::isEnabled(LogLevel::DEBUG, Logger::MAINTENANCE)) {
-            std::stringstream msg;
-            msg << "SynchronizeShard: shortcut worked, done, ";
-            AppendShardInformationToMessage(database, shard, planId, startTime, msg);
-            LOG_TOPIC("f4a5b", DEBUG, Logger::MAINTENANCE) << msg.str();
-          }
-          collection->followers()->setTheLeader(leader);
-          return false;
-        }
-        if (asResult.errorNumber() != TRI_ERROR_REPLICATION_SHARD_NONEMPTY) {
-          // Stop action in this case
-          LOG_TOPIC("daaaa", INFO, Logger::MAINTENANCE) << "SynchronizeShard, error in addFollower (short cut): " << asResult.errorMessage();
-          result(asResult);
-          return false;
-        }
-        // Otherwise move on.
-      } catch (...) {
+    VPackBuilder builder;
+    // build configuration for WAL tailing
+    {
+      {
+        VPackObjectBuilder o(&builder);
+        builder.add(ENDPOINT, VPackValue(ep));
+        builder.add(DATABASE, VPackValue(database));
+        builder.add(COLLECTION, VPackValue(shard));
+        builder.add(LEADER_ID, VPackValue(leader));
+        builder.add("requestTimeout", VPackValue(600.0));
+        builder.add("connectTimeout", VPackValue(30.0));
       }
+      
+      ReplicationApplierConfiguration configuration =
+          ReplicationApplierConfiguration::fromVelocyPack(feature().server(), builder.slice(), database);
+      // will throw if invalid
+      configuration.validate();
```

Comment:
Can we please simply initialize the `ReplicationApplierConfiguration` directly instead of using the velocypack in between? Then we can probably also get rid of this `builder` variable (a few lines down we create yet another `builder` anyway which would shadow this one).
```suggestion
    {
      // build configuration for WAL tailing
      ReplicationApplierConfiguration configuration(feature().server());
      configuration._endpoint = ep;
      configuration._database = database;
      configuration._requestTimeout = 600.0;
      configuration._connectTimeout= 30.0;
      // will throw if invalid
      configuration.validate();
```
As far as I can see `ReplicationApplierConfiguration::fromVelocyPack` does not handle the `COLLECTION` and `LEADER_ID` fields. Is this an oversight?

Reply:
To be honest: I don't like doing this in this PR right now. This is the 3.7 version of a series of changes that we merged for devel, 3.9, 3.8 already. Changing it only in this last open PR wouldn't be ideal, because changes would be lost from 3.8 onwards. So we could open yet more PRs to do these changes in the 3 other branches, which will just delay things further. 
I think code style improvements should be made in devel first (and probably only there) and are not necessarily be done in stable branches.
The collection and leader id attributes will be set somewhere else when calling `syncCollectionCatchup` and `syncCollectionFinalize`. Probably not ideal as it is now, but has been like this for years, and I am not happy with making this change in 3.7 now and then having to re-test everything in all versions.